Reduce Doxygen warnings

johnnonweiler opened the pull request:
Following on from #3115, this addresses some more of the warnings that are produced when running doxygen.  (Note: Unlike the previous PR, these changes are **not** limited to changes to the documentation.  For example, changing int type specifiers to equivalent types to help Doxygen matching things up.)

This PR addresses some of the issues in java_bytecode, goto-programs, util and cbmc.

Please see the individual commits for details.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Move forward declaration of class (to help doxygen)

Before this commit, get_xml_options had the parameters
"(const class xmlt &xml, cmdlinet &cmdline)" in the header file, and
"(const xmlt &xml, cmdlinet &cmdline)" in the cpp file.  Doxygen then
produced a warning "no matching class member found".  This commit
adds a forward declaration "class xmlt;" in the header file, so that
both versions of get_xml_options can have identical parameter lists.

Commit:
Change int type specifiers to equivalent types

Replace 'signed int' with 'int', and 'unsigned int' with 'unsigned',
so that .cpp and .h files are consistent, and doxygen doesn't get
confused

Commit:
Update parameter documentation

Doxygen produces warnings if some parameters are documentated and not
others, or if there is documentation for parameters which don't exist.

Commit:
Move documentation into cpp file

A doxygen \param comment in the header file was referring to the
'goto_model' parameter which was named only in the cpp file, and doxygen
complained that it could not find the parameter.

(Moved other comments for consistency.)

thk123 commented on jbmc/src/java_bytecode/replace_java_nondet.h:
```diff
@@ -19,7 +19,7 @@ class goto_model_functiont;
 /// representation.
 /// \param goto_model: The goto program to modify.
-void replace_java_nondet(goto_modelt &);
+void replace_java_nondet(goto_modelt &goto_model);
```
:pick: Probably the fix is to move the documentation into the cpp file as per coding guidelines.

tautschnig replied:
+1

tautschnig commented on jbmc/src/java_bytecode/java_types.h:
```diff
@@ -695,7 +695,7 @@ inline typet java_type_from_string_with_exception(
 
 /// Get the index in the subtypes array for a given component.
-/// \param t The type we search for the subtypes in.
+/// \param gen_types The subtypes array.
```
Nit-pick: add a colon after `gen_types` as is recommended in the coding guidelines (but arguably it's not done here for the other parameters).

tautschnig commented on src/goto-programs/read_goto_binary.cpp:
```diff
@@ -261,6 +261,8 @@ bool read_object_and_link(
 /// \brief reads an object file, and also updates the config
 /// \param file_name file name of the goto binary
+/// \param dest_symbol_table symbol table to update
+/// \param dest_functions collection of goto functions to update
```
Nit-pick as above: add colons

tautschnig commented on src/util/std_types.h:
```diff
@@ -79,7 +79,7 @@ class symbol_typet:public typet
 /// Check whether a reference to a typet is a \ref symbol_typet.
 /// \param type Source type.
-/// \return True if \param type is a \ref symbol_typet.
+/// \return True if type is a \ref symbol_typet.
```
Ok for correct syntax, but is there some way to make Doxygen understand that this "type" refers to the named parameter "type"? I believe this is what the author intended.

johnnonweiler replied:
I couldn't find a way to explicitly link it to the parameter.  I tried adding emphasis, or marking it as code, but I think it looks best as plain text.

tautschnig replied:
How about using `\p`?

johnnonweiler replied:
I hadn't seen that, thanks.  It certainly looks from the Doxygen documentation like that is the way to go (https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdp).  Their example looks good:

![screen shot 2018-10-15 at 14 25 02](https://user-images.githubusercontent.com/32836213/46953604-364cd480-d086-11e8-8d10-6dbb27e63eee.png)

However, it doesn't look so good in the documentation generated for cbmc:

![screen shot 2018-10-15 at 14 20 51](https://user-images.githubusercontent.com/32836213/46953468-dce4a580-d085-11e8-849d-511c82e73b2e.png)

I guess it's probably best to use the \p in the cbmc documentation now, and then possibly (in future) consider changing the cbmc documentation formatting so that it is more like the standard Doxygen formatting.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 21c7bc9d3ff4fd8ad83d5fe55f41d25c086e5ce0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87929992

johnnonweiler commented:
@tautschnig I have moved the documentation to the cpp file, and added some colons.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: cfe83b97474303f2a8108a89d87e557f65858999).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87952598
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 20a8fe5ec5503181a8c7ec576bd6493c016a8059).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87954380

Commit:
Correct Doxygen syntax

Doxygen was warning that the argument 'type' has muliple @param
documentation sections, and producing incorrect documentation.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 48061b41cfd8e43ba1e8e32b10763093ce9a8e26).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87966511
